text/0000-fabric-private-chaincode.md

```diff
@@ -0,0 +1,284 @@
+ 
+FPC Client SDK extension: This extends the NodeSDK (or Go SDK) with extra functionality that allows users to write end-to-end secure FPC-based applications. In particular, the client SDK extension will provide these core functions: 1) FPC transaction proposal creation (including transparent encryption of arguments) 2) FPC transaction proposal response validation and decryption of the result; 3) FPC deployment/setup functionality. 
+
+FPC SDK: The FPC SDK contains a set of tools to write and build chaincodes. In particular, the SDK contains a build-chain to link a chaincode against our chaincode library and produce a deployment artifact that can be executed by the FPC runtime.
+
+FPC Runtime: The FPC runtime comprises all components that are required to deploy and run an FPC chaincode.
+
+## FPC chaincode deployment
+ 
+The deployment of a FPC chaincode consists of multiple steps; It begins with the standard Fabric 2.0 Lifecycle mechanism to produce a chaincode definition that all participants agree upon. FPC requires adding some additional metadata to the chaincode definition, in particular a cryptographic hash (called the MRENCLAVE in the initial version of FPC) that identifies the code running inside the chaincode enclave. The FPC framework provides tools to generate this MRENCLAVE in a way that enables all parties to screen the code upfront and produce this hash. Note that this is crucial for the security provided by the TEE and is part of the attestation capabilities.
+ 
+In order to support the endorsement model of Fabric, chaincode enclaves executing the same FPC chaincode need access to a set of shared secrets. In particular, these shared or common keys are used to encrypt and decrypt the transaction arguments and state. A key generation and distribution protocol will be part of the FPC deployment process. Note that our current prototype code base does not yet implement the key distribution protocol.  However, the protocol is described in detail and can be found in the FPC repository on github.
+ 
+The Ledger Enclave stores all attestation reports, as signed by the TEE vendor. Reports include information on what chaincode is running; what specific TEE is in use, including version number and hash of the chaincode, and including the public keys of the enclave. With the planned new key management scheme to support multiple endorsing peers, there may be many public keys instead of one. Upon creation each chaincode enclave generates public/private key pairs for signing and for encryption. The public signing key also denotes the chaincode enclave identity. In the new scheme, there will also be a public/private key pair and a symmetric state encryption key per FPC chaincode that are shared among all chaincode enclaves that run the same FPC chaincode using the key distribution protocol above.
+  
+ 
+## FPC Chaincode Lifecycle
```

Comment:
A figure or a flow diagram could help understand the proposed approach better.

Reply:
I agree! Will add this.